socorro/signature/rules.py

```diff
@@ -366,26 +367,37 @@ def generate(self, source_list, crashed_thread=None, delimiter=" | "):
         # have already been normalized at this point.
         new_signature_list = []
         for a_signature in source_list:
-            # If the signature matches the irrelevant signatures regex, skip to the next frame.
+            # If the frame signature is an empty string, skip it
+            if not a_signature:
+                debug_notes.append(f'empty string; ignoring: "{a_signature}"')
+                continue
+
+            # If the frame signature matches the irrelevant signatures regex, skip to
+            # the next frame
             if self.irrelevant_signature_re.match(a_signature):
                 debug_notes.append(f'irrelevant; ignoring: "{a_signature}"')
                 continue
 
-            # If the frame signature is a dll, remove the @xxxxx part.
-            if (
-                not a_signature.startswith("(unloaded")
-                and ".dll" in a_signature.lower()
+            # Skip repeating dll/so signatures
+            if not a_signature.startswith("(unloaded") and (
+                # A library on Windows or Linux
+                a_signature.lower().endswith((".dll", ".so"))
+                # XUL library on macOS
+                or a_signature == "XUL"
+                # Case where symbols file has "name omitted" as function symbol
+                or a_signature == "<name omitted>"
+                # Case where symbols file only has public function names. bug
+                # #1685178
+                or a_signature.startswith("<unknown in")
```

Comment:
This adds a few more cases to collapse plus it adds some explanation of each case.

Reply:
Doing all of this with string matching looks kind of brittle to me. It feels like the processing should rather be done on typed and structured data, and only at the end we convert that to a string. (I no a change like this is totally out of scope for this PR, so I'm not asking for any change here.)

Reply:
Currently, all things symbols is done in strings. The strings come from multiple places:

1. compiler generated symbols from builds Mozilla does and builds done by Microsoft, Apple, Nvidia, various Linux distributions, etc
2. compiler symbols that dump_syms fixes in its demangling before putting them in symbols files
3. changes that the stackwalker makes to symbols before emitting the stackwalker output when processing a minidump
4. changes we make to normalize symbols in signature generation

It could be easier if the data were more structured, but getting from here to there feels like an insurmountable task across a bunch of projects, organizations, and decades of software development history.

Reply:
I wasn't really thinking of the symbols themselves, but rather of the frame signatures generated by `normalize_frame()`. Instead of being just strings, they could be `Function(name, line)`, `Module(name)`, `UnloadedModule(name)`, `Offset(offset)` or similar. Then instead of `a_signature.startswith("(unloaded")` we could do something more like `isinstance(a_signature, UnloadedModule)`.

I didn't really put much thought into this comment, and I'm only looking at a tiny piece of this whole system without having much experience with the rest, but sometimes a fresh perspective is useful. Often, it isn't.